JoomGalleryfriends/JG4-dev
# [Frontend] Router

Adding the router to the frontend. Making the URL's as close as possible to the ones of the core (com_content, com_users, ...). But having the possibility to switch to the old JG3 router for compatibility.

The new, modern Router (RouterView, since Joomla 3.8) was implemented. This changes quite a lot in contrast to the legacy router of the old JoomGallery. In order for the Router to work correctly, JoomGallery will create three Menu Items by default during installation. The main view where all the routing is attached to is the gallery view. The gallery view is implemented only very rough in this PR. Maybe we could add a nicer template to this view after merging this PR?

## How the router works
The modern router creates URLs based on found menuitems and creates the URL segments based on the hirarchy of the found menuitems. If there are any query parameters left which can not be translated with the available menuitems then they will be attached at the end of the SEF URL.

**Available menuitem types**
- Gallery view (root view)
- List of categories (root view)
- Category view (child of gallery view)
- List of images (child of gallery view)
- Image view (child of list of images)

**Example:**
`index.php/gallery/images/1-testimage.html?catid=1`
|-> `gallery` is the alias of a menuitem of type gallery
|-> `images` is the alias of a menuitem of type list of images
|-> `1-testimage` is the ID and alias of the image to be displayed
|-> `html` is the format to be displayed. Html means that the detail view of the image is displayed.
|-> `catid=1` there was a query parameter `catid` which could not be translated to a segment. This query parameter will be added at the end of the SEF URL.

## How to test this PR
Activate the SEF routing in the Joomla global configuration. Add content to the component and create different menuitems. Then flip through the component views in the frontend and check if
- No 404 (view not found) appears
- URL looks as if you would expect them knowing the core URL behavior and the description provided above
